### ISSUE TITLE
[TPE] Premature region abandonment due to unfair co-parameter  evaluation in parallel optimization

### ISSUE THREAD

**Issue by aditirani** [optuna/optuna#6623]
## Motivation

While investigating how Optuna handles parallel trial execution (`n_jobs > 1`), 
I traced the internals of `study.optimize()` down to the TPE sampler and identified 
a systematic bias in how parameter regions get classified as "good" or "bad".

The core issue: `_split_complete_trials_single_objective()` classifies trials 
into `below` (good) and `above` (bad) purely based on objective value. This means 
a parameter region can get permanently condemned based on a single trial where 
*co-parameters* — not the parameter of interest — were responsible for the poor result.

This bias is amplified under parallel execution because when `n_jobs=2`, 
Trial-1 and Trial-2 are suggested before either has a result. The sampler 
suggests both trials with no information about each other. If Trial-1 happens 
to pair a promising region (e.g. `n_estimators=1000`) with a bad co-parameter 
(e.g. `learning_rate=0.2`), the entire `n_estimators=1000` region gets labeled 
bad — even though the region itself was never fairly evaluated.

## Example

Consider tuning XGBoost with `n_jobs=2`:

```python
import optuna
import numpy as np

def objective(trial):
    n_estimators = trial.suggest_int("n_estimators", 100, 2000)
    learning_rate = trial.suggest_float("learning_rate", 0.001, 0.5, log=True)
    max_depth = trial.suggest_int("max_depth", 3, 10)
    
    # Simulate: n_estimators=1000 is only good when learning_rate is low
    # This mirrors real XGBoost behavior
    if n_estimators > 1200 and learning_rate < 0.05:
        return -0.95   # global optimum — requires correct co-parameters
    elif n_estimators < 500:
        return -0.75   # local optimum — easy to find, TPE gets stuck here
    else:
        return -0.55 + np.random.normal(0, 0.02)

study = optuna.create_study(direction="minimize")
study.optimize(objective, n_trials=100, n_jobs=2)
```

What happens step by step:

Trial-0: n_estimators=1000, learning_rate=0.2  → value=-0.55  (bad)
Trial-1: n_estimators=300,  learning_rate=0.1  → value=-0.75  (good)
After these two trials, TPE's split:
below (good): Trial-1 → n_estimators ~ 300 region gets reinforced
above (bad):  Trial-0 → n_estimators ~ 1000 region gets condemned
From trial 10 onward, TPE samples almost exclusively around n_estimators ~ 300.
Ground truth that TPE never discovers:
n_estimators=1000 + learning_rate=0.01 → value=-0.95  (global optimum)

The `n_estimators=1000` region was condemned because it was evaluated 
alongside `learning_rate=0.2`. It was never tried with a low learning rate.
TPE had no mechanism to distinguish "this region is bad" from 
"this region was evaluated unfairly".

---

## Root Cause in the Source Code

I traced this to `_split_complete_trials_single_objective()` in 
`optuna/samplers/_tpe/sampler.py`:

```python
# optuna/samplers/_tpe/sampler.py (~line 735)

def _split_complete_trials_single_objective(
    trials: Sequence[FrozenTrial], study: Study, n_below: int
) -> tuple[list[FrozenTrial], list[FrozenTrial]]:
    if study.direction == StudyDirection.MINIMIZE:
        sorted_trials = sorted(trials, key=lambda trial: cast(float, trial.value))
    else:
        sorted_trials = sorted(trials, key=lambda trial: cast(float, trial.value), reverse=True)

    # Purely value-based split
    # No awareness of whether bad trials were fairly evaluated
    return sorted_trials[:n_below], sorted_trials[n_below:]
```

The call chain that leads here is:
study.optimize(n_jobs=2)
└── _optimize()
└── _optimize_parallel()        [ThreadPoolExecutor]
└── _optimize_sequential()
└── _run_trial()
└── study.ask()
└── TPESampler._sample()
└── _split_trials()
└── _split_complete_trials()
└── _split_complete_trials_single_objective()
← problem is here

---

## Why Existing Solutions Do Not Fully Address This

| Existing approach | What it does | Why it doesn't fix this |
|---|---|---|
| `multivariate=True` | Models joint param distribution in Parzen estimator | Only affects `_build_parzen_estimator()` — has zero effect on the good/bad split which happens upstream in `_split_complete_trials_single_objective()` |
| `constant_liar=True` | Assigns worst value to RUNNING trials | Addresses parallel suggestion overlap — does not address unfair co-param evaluation of already COMPLETE trials |
| `n_startup_trials` | Random exploration at start | Reduces early bias but does not prevent mid-optimization abandonment |
| CMA-ES sampler | Full covariance matrix modeling | Different sampler entirely — not a fix within TPE |

I specifically verified that `multivariate=True` does not affect the split 
by checking the source — the `multivariate` flag is only read inside 
`_build_parzen_estimator()`, which is called *after* the split is already done.

---

## Proposed Solution

Add an optional `coparameter_fairness` flag to `TPESampler` that checks, 
before finalizing a trial as "above" (bad), whether that trial's co-parameters 
were similar to those of the current good trials. If not — the trial was 
unfairly evaluated and should be kept in the exploration pool rather than 
condemned.

```python
# Modified _split_complete_trials_single_objective()

def _split_complete_trials_single_objective(
    trials: Sequence[FrozenTrial],
    study: Study,
    n_below: int,
    coparameter_fairness: bool = False,    # new — default False for backward compat
    fairness_threshold: float = 0.5,       # new — similarity threshold
) -> tuple[list[FrozenTrial], list[FrozenTrial]]:

    if study.direction == StudyDirection.MINIMIZE:
        sorted_trials = sorted(trials, key=lambda trial: cast(float, trial.value))
    else:
        sorted_trials = sorted(trials, key=lambda trial: cast(float, trial.value), reverse=True)

    below = sorted_trials[:n_below]
    above = sorted_trials[n_below:]

    if not coparameter_fairness or len(below) == 0:
        return below, above    # original behavior fully preserved

    # Co-parameter fairness check
    good_param_sets = [t.params for t in below]
    fairly_bad = []
    unfairly_bad = []    # these were never tried with good co-params

    for trial in above:
        if _was_fairly_evaluated(trial, good_param_sets, fairness_threshold):
            fairly_bad.append(trial)
        else:
            # Keep in exploration pool — do not condemn yet
            unfairly_bad.append(trial)

    return below + unfairly_bad, fairly_bad
```

Proposed API:

```python
sampler = optuna.samplers.TPESampler(
    coparameter_fairness=True,   # opt-in
    fairness_threshold=0.5       # tunable
)
study = optuna.create_study(sampler=sampler, direction="maximize")
study.optimize(objective, n_trials=100, n_jobs=2)
```

Default is `False` — zero change in behavior for existing users.

---

## Impact

This problem is most severe when all of the following are true:

- `n_jobs > 1` — parallel trials share no result information during execution
- High-dimensional search space — more co-parameters to interact
- Strong parameter interactions — e.g. `n_estimators` + `learning_rate` 
  in gradient boosting, `hidden_size` + `dropout` in neural networks
- Early random trials happen to pair a promising parameter value 
  with bad co-parameters — which is statistically likely in high dimensions

---

## Questions for Maintainers

1. Is `_split_complete_trials_single_objective()` the right layer to 
   introduce this check, or would it be better placed directly in `_split_trials()`?
2. Should `fairness_threshold` be adaptive — decreasing as the trial budget 
   is consumed — rather than a fixed value?
3. The current proposal uses cosine similarity over raw parameter values. 
   Would it be more principled to use the normalized internal representations 
   that Optuna already computes for the Parzen estimator?
4. Would you welcome a PR implementing this? I am happy to write the full 
   implementation, tests, and documentation.

---

## Environment

- Optuna version: latest master
- Python: 3.12
- Primary file: `optuna/samplers/_tpe/sampler.py`
- Primary function: `_split_complete_trials_single_objective()` (~line 735)
- Related function: `_split_trials()` (~line 686)

**Comment by nabenabe0928** [comment]
Hey, thank you for the opinion.
Long story short, I understand what you mean (and I even published a paper that partially addresses your point -- the PED-ANOVA paper), but I'd say we opt to keep the current API. That said, you could contribute to OptunaHub for this feature.

---

First of all, I need to say that the current split algorithm is fully intended and it's simple yet effective in practice.

> If Trial-1 happens to pair a promising region (e.g. n_estimators=1000) with a bad co-parameter (e.g. learning_rate=0.2), the entire n_estimators=1000 region gets labeled bad — even though the region itself was never fairly evaluated.

This isn't correct for `multivariate=True`.
It partially affects the score, though your statement is true for `multivariate=False`.

Also, it's worth noting that we just say "bad" for most observations, and only the top 10% is good.
At the end of the day, there is no way to know that (your) underestimated Parameters are really underestimated or not without observing.

It's kinda true that the unfairly underestimated Parameters may be important at the late stage of optimization (this is what the PED-ANOVA's local importance is for), but it's just a trade-off between sample efficiency and conservativism. We first need to identify the most influential parameter, and need to fix it (or narrow down the parameter range). Your tweak wants to be accounted only after this. It's inefficient to apply the tweak before we are sure the important parameters. But it's hard to say that the bad performance didn't stem from the "unfairly underestimated" Parameters, justifying the current split algorithm.

The good part of TPE is basically the rough sample efficiency. Namely, we give up precise statistical modeling and focus only on producing compromisable trials with a limited budget.
By introducing the co-parameter feature, the good part is, in principle, cancelled out more or less.

That said, its worth trying from the research perspective (i'd say).

**Comment by y0z** [comment]
As there are no plans to address this at present, I am closing this issue.
